File: LeetCode/101-200/105. 从前序与中序遍历序列构造二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer,Integer> map;
    int[] preorder;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0) return null;
        this.preorder  =preorder;
        map = new HashMap<>();
        for (int i = 0; i < preorder.length; i++) {
            map.put(inorder[i],i);
        }
         return build(0, preorder.length -1,0, inorder.length - 1);
    }

    public TreeNode build(int p_start,int p_end, int i_start, int i_end) {
        if (p_start > p_end) return  null;
        int rootV = preorder[p_start];
        TreeNode root = new TreeNode(rootV);
        Integer mid = map.get(rootV);
        int left = mid - i_start;
        root.left = build(p_start + 1,p_start + left,i_start,mid - 1);
        root.right = build( p_start + left + 1,p_end,mid + 1,i_end);
        return  root;
    }
}

File: LeetCode/101-200/105.从前序与中序遍历序列构造二叉树.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func buildTree(preorder []int, inorder []int) *TreeNode {
	m := map[int]int{}
	for i, v := range inorder {
		m[v] = i
	}
	var build func(int, int) *TreeNode
	build = func(l int, r int) *TreeNode {
		if l > r {
			return nil
		}
		val := preorder[0]
		preorder = preorder[1:]
		
		root := &TreeNode{Val: val}
		mid := m[val]
		root.Left = build(l, mid-1)
		root.Right = build(mid+1, r)
		return root
	}
	return build(0, len(preorder)-1)
}